### PR TITLE
feat(tofs): use TOFS to distribute configuration templates to minion

### PR DESCRIPTION
**Pull request by baby-gnu**
BREAKING CHANGES: the templates are now under “libvirt/server/config/files/”.

* docs/TOFS_pattern.rst: document the TOFS pattern.

* libvirt/libtofs.jinja: TOFS library.

* libvirt/server/config/files.sls: lookup configuration templates with
  “files_switch” macro.

* pillar.example: provide commented “tofs” pillar configuration as
  example.

### CODE REVIEW

File: pillar.example [hunk 1]
```diff
@@ -19,3 +19,32 @@ libvirt:
...
+      - any/path/can/be/used/here
+      - id
+      - roles
+      - osfinger
+      - os
+      - os_family
+
+    # All aspects of path/file resolution are customisable using the options below.
+    # This is unnecessary in most cases; there are sensible defaults.
+    # path_prefix: template_alt
```
Comment (myii) [inline pillar.example]: @baby-gnu I recently added a couple of lines here in the `template-formula`, based upon a PR we received in the `nginx-formula`.

Comment (myii) [reply pillar.example]: Sorry, it's just above this line (on my mobile device, hit the wrong line!).

Comment (baby-gnu) [reply pillar.example]: I just repushed a fixed version.

Thanks.

**Comment by baby-gnu** [comment]
@myii the problem is that now I need to use a custom pillar.example to avoid jinja rendering error ;-)

**Comment by myii** [comment]
@baby-gnu Sorry about that!  The comment indenting in the `template-formula` and `nginx-formula` work, so they pass the `yamllint`.  I've actually got `yamllint` integrated into my editor, so I catch the errors before I push.  Thanks for going through all of the trouble.

**Comment by baby-gnu** [comment]
> @baby-gnu Sorry about that! The comment indenting in the `template-formula` and `nginx-formula` work, so they pass the `yamllint`. I've actually got `yamllint` integrated into my editor, so I catch the errors before I push. Thanks for going through all of the trouble.

We should be very careful with the addition of the following comments in `pillar.example`

```yaml
    # Default path: salt://{{ path_prefix }}/{{ dirs.files }}/{{ dirs.default }}
    #         I.e.: salt://template/files/default
```

It must be done only on formula using a dedicated pillar for tests like [nginx-formula](https://github.com/saltstack-formulas/nginx-formula/blob/17291a0ae2c2554707b79d897bb6ddec716e8426/kitchen.yml#L208) or salt will try to compute `{{ path_prefix }`, `{{ dirs.files }}` and `{{ dirs.default }}` when rendering the pillars.

I [did it](https://github.com/saltstack-formulas/libvirt-formula/commit/6cae6455f2ea5912e88098300e67043a4bdc8b97#diff-fdbca9ed07a634829a3548f874dc925a)  for that formula.

**Comment by myii** [comment]
@baby-gnu Good catch, that should be fixed in all of the formulas so far.  We shouldn't be forced to avoid using `pillar.example` in Kitchen.  So what do you think about the following instead?

```diff
-    # Default path: salt://{{ path_prefix }}/{{ dirs.files }}/{{ dirs.default }}
+    # Default path: salt://< path_prefix >/< dirs.files >/< dirs.default >
```

Please mention if you have a better suggestion.

**Comment by baby-gnu** [comment]
@myii I completely agree, I'm pushing the fix with your suggestion right now.

**Comment by myii** [comment]
@baby-gnu Thanks, I'll push that to the other formulas.

**Comment by myii** [comment]
@baby-gnu I'm a little busy right now but I'll test this PR in a few hours and then merge it if all is OK.

**Comment by myii** [comment]
@baby-gnu Comparing debug output before and after this PR (map and `files_switch` x2):

```diff
--- before
+++ after
@@ -1,4 +1,4 @@
-https://travis-ci.org/myii/libvirt-formula/jobs/593221139#L1110-L1134
+https://travis-ci.org/myii/libvirt-formula/jobs/593221266#L1298-L1335
        [DEBUG   ] Map   : .../libvirt-formula/.kitchen/yaml_dump/Debian-10-2019.2.0-n/a-e683178-py3.yaml
        daemon_config_path: /etc/default
        extra_pkgs:
@@ -24,24 +24,120 @@
        python2_pkg: python-libvirt
        python3_pkg: python3-libvirt
        qemu_pkg: qemu-kvm
+       tofs:
+         files_switch:
+         - any/path/can/be/used/here
+         - id
+         - roles
+         - osfinger
+         - os
+         - os_family
+         source_files:
+           libvirt-server-config-files-libvirtd-file-managed:
+           - libvirtd_alt.conf.jinja
+           libvirt-server-config-files-daemonconfig-file-managed:
+           - daemon_libvirtd_alt.jinja
 ---
-https://travis-ci.org/myii/libvirt-formula/jobs/593221139#L1294-L1302
+https://travis-ci.org/myii/libvirt-formula/jobs/593221266#L1548-L1598
        libvirt-server-config-files-libvirtd-file-managed:
          file.managed:
            - name: /etc/libvirt/libvirtd.conf
            - template: jinja
-           - source: salt://libvirt/server/config/templates/libvirtd.conf.jinja
+           - source: 
+             - salt://libvirt/server/config/files/any/path/can/be/used/here/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/any/path/can/be/used/here/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/44f1f6eb077e/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/44f1f6eb077e/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/roles/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/roles/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/Debian-10/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/Debian-10/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/server/config/files/default/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/config/files/default/libvirtd.conf.jinja
+             - salt://libvirt/server/files/any/path/can/be/used/here/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/any/path/can/be/used/here/libvirtd.conf.jinja
+             - salt://libvirt/server/files/44f1f6eb077e/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/44f1f6eb077e/libvirtd.conf.jinja
+             - salt://libvirt/server/files/roles/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/roles/libvirtd.conf.jinja
+             - salt://libvirt/server/files/Debian-10/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/Debian-10/libvirtd.conf.jinja
+             - salt://libvirt/server/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/server/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/server/files/default/libvirtd_alt.conf.jinja
+             - salt://libvirt/server/files/default/libvirtd.conf.jinja
+             - salt://libvirt/files/any/path/can/be/used/here/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/any/path/can/be/used/here/libvirtd.conf.jinja
+             - salt://libvirt/files/44f1f6eb077e/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/44f1f6eb077e/libvirtd.conf.jinja
+             - salt://libvirt/files/roles/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/roles/libvirtd.conf.jinja
+             - salt://libvirt/files/Debian-10/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/Debian-10/libvirtd.conf.jinja
+             - salt://libvirt/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/files/Debian/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/Debian/libvirtd.conf.jinja
+             - salt://libvirt/files/default/libvirtd_alt.conf.jinja
+             - salt://libvirt/files/default/libvirtd.conf.jinja
            - require:
              - pkg: libvirt-server-package-install-pkg-installed
            - watch_in:
              - service: libvirt-server-service-running-service-running
 ---
-https://travis-ci.org/myii/libvirt-formula/jobs/593221139#L1304-L1312
+https://travis-ci.org/myii/libvirt-formula/jobs/593221266#L1600-L1650
        libvirt-server-config-files-daemonconfig-file-managed:
          file.managed:
            - name: /etc/default/libvirtd
            - template: jinja
-           - source: salt://libvirt/server/config/templates/Debian/daemon_libvirtd.jinja
+           - source: 
+             - salt://libvirt/server/config/files/any/path/can/be/used/here/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/any/path/can/be/used/here/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/44f1f6eb077e/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/44f1f6eb077e/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/roles/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/Debian-10/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/Debian-10/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/server/config/files/default/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/config/files/default/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/any/path/can/be/used/here/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/any/path/can/be/used/here/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/44f1f6eb077e/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/44f1f6eb077e/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/roles/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/roles/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/Debian-10/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/Debian-10/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/server/files/default/daemon_libvirtd_alt.jinja
+             - salt://libvirt/server/files/default/daemon_libvirtd.jinja
+             - salt://libvirt/files/any/path/can/be/used/here/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/any/path/can/be/used/here/daemon_libvirtd.jinja
+             - salt://libvirt/files/44f1f6eb077e/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/44f1f6eb077e/daemon_libvirtd.jinja
+             - salt://libvirt/files/roles/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/roles/daemon_libvirtd.jinja
+             - salt://libvirt/files/Debian-10/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/Debian-10/daemon_libvirtd.jinja
+             - salt://libvirt/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/files/Debian/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/Debian/daemon_libvirtd.jinja
+             - salt://libvirt/files/default/daemon_libvirtd_alt.jinja
+             - salt://libvirt/files/default/daemon_libvirtd.jinja
            - require:
              - pkg: libvirt-server-package-install-pkg-installed
            - watch_in:
```

* Everything working as expected.
* For a further PR, maybe it's worth modifying `pillar.example` so all of the extra entries aren't produced under `source`.

**Comment by myii** [comment]
Merged, thanks @baby-gnu!

**Comment by myii** [comment]
@baby-gnu Just noticed that this didn't trigger a major release.  The situation is that you used `BREAKING CHANGES` in the commit message instead of `BREAKING CHANGE`.  I don't think it's a big deal but you're welcome to resolve it if you think it's important.

**Comment by baby-gnu** [comment]
@myii thanks, I always make that mistake ;-)

**Comment by myii** [comment]
@baby-gnu No problem, why do you think I avoid it!  I _still_ have to keep checking the `CONTRIBUTING` docs!

**Pull request merged**